### ISSUE TITLE
Implementation for neosr/BasicSR

### ISSUE THREAD

**Issue by neosr-project** [2minkyulee/Noise-free-Optimization-in-Early-Training-Steps-for-Image-Super-Resolution#3]
Hi, I tried to port ECO to [neosr](https://github.com/muslll/neosr)/BasicSR. However, due to how the main training loop happens, it seems like the original sample code provided in this repo needs to be modified. Could any of the authors verify if this is correct?

```python

    def init_training_settings(self):
        # flag to enable ECO optimization:
        self.eco = self.opt["train"].get("eco", False)
        # ECO amount of iters, defaults to 125k like in official
        self.eco_iters = self.opt["train"].get("eco_iters", 125000)
        # ECO init iters
        self.eco_init = self.opt["train"].get("eco_init", 15000)

        # self.lq are the LQ images, size 1/upscale-ratio
        # self.gt are the GT images
        # self.net_g is the generator network
        # self.pretrain is the path to a pretrained generator

    def eco_strategy(self, current_iter):
        """ Adapted version of "Empirical Centroid-oriented Optimization":
            https://arxiv.org/abs/2312.17526
        """

        with torch.no_grad():
            # define alpha with sigmoid-like curve, slope/skew at 0.25
            a = 1 / (1 + math.exp(-1 * (10 * (current_iter / self.eco_iters - 0.25))))
            # network prediction
            self.net_output = self.net_g(self.lq)
            # define gt centroid
            self.gt = ((1 - a) * self.net_output) + (a * self.gt)
            # downsampled prediction
            self.lq_scaled = torch.clamp(
                F.interpolate(
                    self.net_output,
                    scale_factor=1 / self.scale,
                    mode="bicubic",
                    antialias=True,
                ),
                0,
                1,
            )
            # define lq centroid
            self.output = ((1 - a) * self.lq_scaled) + (a * self.lq)

        # predict from lq centroid
        self.output = self.net_g(self.output)

        return self.output, self.gt

    def optimize_parameters(self, current_iter):

            # eco
            if self.eco and current_iter <= self.eco_iters:
                if current_iter < self.eco_init and self.pretrain is None:
                    self.output = self.net_g(self.lq)
                else:
                    self.output, self.gt = self.eco_strategy(current_iter)
            else:
                self.output = self.net_g(self.lq)

            # compute losses based on self.output and self.gt
            l_g_total = 0
            loss_dict = OrderedDict()
            # L1 loss
            if self.cri_pix:
                l_g_pix = self.cri_pix(self.output, self.gt)
                l_g_total += l_g_pix
            loss_dict['l_g_pix'] = l_g_pix
            loss_dict['l_g_total'] = l_g_total
            # backward
            self.gradscaler.scale(l_g_total).backward()
            # optimizer step
            self.gradscaler.step(self.optimizer_g)
            # zero grads
            self.optimizer_g.zero_grad(set_to_none=True)
```

The LQ images in neosr/basicsr come at the size 1/upscale-ratio. So in the `eco_strategy` It has to blend the downsampled LQ and original LQ, and then pass through the generator network to upsample, otherwise the losses can't be computed due to tensors having different sizes.

**Comment by neosr-project** [comment]
Here is a visualization of `self.gt` and `self.output` (respectively), click to enlarge:

![eco_gt](https://github.com/2minkyulee/Noise-free-Optimization-in-Early-Training-Steps-for-Image-Super-Resolution/assets/132400428/0e4c6e03-e995-411d-a81a-bba4b0289c54)
![eco_output](https://github.com/2minkyulee/Noise-free-Optimization-in-Early-Training-Steps-for-Image-Super-Resolution/assets/132400428/a99c1bcd-27bf-4b3b-8e7d-65b2c671a90e)

I also took the liberty to modify the growth of `alpha`. The official implementation seems to grow it linearly:

```
a = min(current_iter / self.eco_iters, 1.0)
```

```mermaid
xychart-beta
    title "Linear alpha"
    x-axis [1, 25, 50, 75, 99]
    y-axis "alpha"
   line [0.01, 0.25, 0.5, 0.75, 0.99]
```

I modified it to be a sigmoid-like curve, with slope/skew at 0.25. This means at 25% of total_iters it will achieve a=0.5, and then increase it smoothly up to 1.0:

```
a = 1 / (1 + math.exp(-1 * (10 * (current_iter / self.eco_iters - 0.25))))
```

```mermaid
xychart-beta
    title "Sigmoid alpha"
    x-axis [1, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 99]
    y-axis "alpha"
   line [0.08317269649392238, 0.11920292202211755, 0.18242552380635635, 0.2689414213699951, 0.3775406687981454, 0.5, 0.6224593312018545, 0.7310585786300049, 0.8175744761936437, 0.8807970779778823, 0.9241418199787566, 0.9525741268224334, 0.9706877692486436, 0.9820137900379085, 0.9890130573694068, 0.9933071490757153, 0.995929862284104, 0.9975273768433653, 0.998498817743263, 0.9993891206405656]
```

**Comment by 2minkyulee** [comment]
### Codes

As you have mentioned in the comment, ECO requires a pretrained network. However, it should be an independent and frozen network w.r.t the actual network we aim to train. In terms of our paper, f and f_hat should be independent networks where f_hat is frozen. 

We have tried the following settings but the training collapsed.
Instead of using a fully trained and frozen f_hat, 
1) training f_hat from scratch simultaneously with f 
2) f_hat as an EMA version of f.


In the code you have provided, it seems that net_g acts as both f and f_hat. We suggest the followings.
1) declare an additional network with fully pretrained weights.
2) disable gradient calculations (both _torch.no_grad()_ and _requires_grad=False_) except for the actual training step. Disable grad calc in _eco_strategy_ in your code.
3) eco_iters should be identical to the total training iter. Note that 125K is chosen since we use x8 larger mini-batch size, and accordingly a larger learning rate than general training setups. (125K * 8  = 1000K)

**Additionally, make sure that the downsampling method to generate lq_scaled and the test LR images are identical. If they are different, we expect performance drop.**


### Alpha scheduling (Mixup)
The code snippet in our readme is a simple version of the actual implementation. This is because we didn't observe significant performance changes over different scheduling (as long as it gradually changes from 0 to 1). If you want to know the actual implementation, refer to the following. This code is the scheduling scheme in our actual experiments, which are simple heuristics to gradually change alpha. Others should work well too.

https://github.com/2minkyulee/Noise-free-Optimization-in-Early-Training-Steps-for-Image-Super-Resolution/blob/46f2b12a4111dabfd04af633a892764e2aab8ffe/ECO/trainer/trainer.py#L209

**Comment by 2minkyulee** [comment]
When implementing on basicsr, do the following.

1) Add some code so that the class SRModel can track alpha. 
(Refer to https://github.com/XPixelGroup/BasicSR/blob/master/basicsr/models/sr_model.py)

2) Define a dataset that reads 4 images (lr_original, hr_origianl, lr_pretrained, sr_pretrained) and register it w/ @DATASET_REGISTRY.register() 
(Refer to basicsr/data/paired_image_dataset.py)

3) Modify feed_data of SRModel and set self.lq / self.gt as the alpha-blended images.

If you want to keep the original SRModel as it is, define ECOModel that inherits SRModel, and register it appropriately. Then, override only specific parts mentioned above.

**Comment by neosr-project** [comment]
> When implementing on basicsr, do the following. [...]

Yes, that approach works if you downsample the images prior to training (such as with matlab bicubic). However the code above does it on-the-fly (at runtime), using [torch's interpolate](https://pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html).
What I was asking about is if the implementation proposed above is correct to what the paper describes, without considering prior-downsampling. Thanks in advance.

**Comment by neosr-project** [comment]
> In the code you have provided, it seems that net_g acts as both f and f_hat. 

Yes, that is right. However, the user can specify a frozen pretrained network, which `self.net_g` inherits.

> 2. disable gradient calculations (both _torch.no_grad()_ and _requires_grad=False_) except for the actual training step. Disable grad calc in _eco_strategy_ in your code.

I tried using `@torch.no_grad()` decorator on `eco_strategy`, however it produces wrong outputs. I'll try again using a different method.

> **Additionally, make sure that the downsampling method to generate lq_scaled and the test LR images are identical. If they are different, we expect performance drop.**

Yes, like mentioned, it is currently using torch's interpolate with bicubic.

> we didn't observe significant performance changes over different scheduling (as long as it gradually changes from 0 to 1). 

I see. Thanks!

**Comment by neosr-project** [comment]
> In terms of our paper, f and f_hat should be independent networks where f_hat is frozen.

I think I misunderstood this comment, sorry. So the inference on LQ must be done by a fully frozen network? The code provided in the begining is training net_g and using that same in-training network to do the inference on LQ/GT. Have you or any of the authors tried such setting? My intuition would be that since the network is still learning, it will improve over a frozen network, and thus making better results as iters process. I haven't properly tested this hypothesis, however the code seems to work if the pretrained network is provided.

**Comment by 2minkyulee** [comment]
> So the inference on LQ must be done by a fully frozen network?

I'm confused about the term 'inference on LQ'. 
If you mean 'inference on LQ in **order to generate sr_pretrained and lr_pretrained** (lq_scaled in your code),
yes, this should be done on a pretrained and frozen network.

> training net_g and using that same in-training network to do the inference on LQ/GT

I believe that a shortcut exists, and the network may potentially collapse. When the network is shared for both phases, the network may simply generate completely black images.  Please refer to the "**We have tried the following settings**" part of my previous comment.

> My intuition would be that since the network is still learning, it will improve over a frozen network

I agree that further finetuning might help, instead of freezing it. However, fine-tuning should be based on original (real) image pairs, NOT the in-training network, since it may collapse (please refer above). 


> I tried using @torch.no_grad() decorator on eco_strategy, however it produces wrong outputs.

Regarding that all test/val codes (my version, basicsr, kair, and more) disable gradient,  this shouldn't affect the output values. Please double-check for any potential bugs.

#### Regarding the resize function
To the best of my knowledge, there are several variants of the bicubic resizing function (cv2, PIL, matlab ... etc.).
Despite having the identical name "bicubic interpolation", results may vary since they have slightly different parameters. 
Accordingly, using pytorch bicubic resize might not match the original LR generation pipeline. 
If you want to stick with the pytorch resize for simplicity, make sure to set ANTIALIAS option to true. (MATLAB bicubic resizing uses antialiasing as default). As an alternative, consider using BasicSR's codes that reimplemented MATLAB bicubic function in Python. Also, outputs of bicubic resize are not guaranteed to be in the 0~1 range, thus, it is better to manually clamp them.
(Refer to the following for the Python version MATLAB bicubic resize function. https://github.com/XPixelGroup/BasicSR/blob/master/basicsr/utils/matlab_functions.py)


However, as training proceeds, the network might be able to adapt through slight discrepancies that come from different resize functions. That is, simply using pytorch resize might be fine. This is because when alpha reaches 1, it only uses the original image pairs. But please note that we haven't tested this setting.

**Comment by neosr-project** [comment]
> If you mean 'inference on LQ in **order to generate sr_pretrained and lr_pretrained** (lq_scaled in your code), yes, this should be done on a pretrained and frozen network.

Yes, that's what I mean. Thanks.

> When the network is shared for both phases, the network may simply generate completely black images.

Yes, that does seem to be the case, because the network has not yet learned enough features, so the centroid (which takes the inference of that network) gets messed up. However, when you start training from a pretrained network, the network does not collapse, as the inference to generate the centroid is from a network that has already learned such basic image features.

> Please double-check for any potential bugs.

Will do, thanks.

> If you want to stick with the pytorch resize for simplicity, make sure to set ANTIALIAS option to true.
> outputs of bicubic resize are not guaranteed to be in the 0~1 range, thus, it is better to manually clamp them.

Interesting. I'll do that and update the code on the first comment for future reference, thanks again :+1:

**Comment by himkmk** [comment]
👍
